hazelcast-sql/src/main/java/com/hazelcast/sql/impl/type/QueryDataType.java

```diff
@@ -148,6 +177,30 @@ public TypeKind getObjectTypeKind() {
         return objectTypeKind;
     }
 
+    public void addField(String name, QueryDataType type) {
+        if (objectFieldsBuilder == null) {
+            objectFieldsBuilder = ImmutableList.builder();
+        }
+        objectFieldsBuilder.add(new QueryDataTypeField(name, type));
+    }
+
+    public void finalizeFields() {
+        if (objectFields == null) {
+            objectFields = objectFieldsBuilder.build();
+            objectFieldsBuilder = null;
```

Comment:
after resetting the builder I can call addField again (a new builder will be created)

Reply:
[`8818039c`..`217f6dc0`][1]

[1]: https://github.com/hazelcast/hazelcast/compare/8818039c4cdcb25814b2d161a97c8032d229f536..217f6dc09634511c35cda5fc73dba20851abd051

Reply:
please add tests for add, finalize, get methods

Reply:
They are already covered in `testSerialization()`, `testDigestEscaping()` and `testFinalization()`. `QueryDataTypeTest` tests functionalities, not methods.

Reply:
Please take a look at these tests. When I asked to add tests I meant to cover all possible orders of calling get, add, and finalize to be sure that these methods work as expected. Some cases were missed in your tests.
Please notice that the behavior of HazelcastObjectType and QueryDataType are different, for example when trying to get empty fields. Just to be sure that this is okey.

https://github.com/shultseva/hazelcast/blob/feature/querydatatype-enum-props/hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/validate/types/HazelcastObjectTypeTest.java#L57

https://github.com/shultseva/hazelcast/blob/feature/querydatatype-enum-props/hazelcast-sql/src/test/java/com/hazelcast/sql/impl/type/QueryDataTypeTest.java#L247

Reply:
1. `QueryDataType` is created when parsing the query and enriched during field validation, whereas `HazelcastObjectType` is created during query execution. So, `QueryDataType` is just a placeholder initially, but `HazelcastObjectType` is always expected to have fields. That's why we don't need to test calling `finalizeFields()`/`getFieldList()` first for `HazelcastObjectType`.
2. `QueryDataType` and `HazelcastObjectType` are not general utilities and they are not exposed to users. That's why they don't need to be tested with complete path coverage. `testFinalization()` is designed to have complete branch coverage, which should be enough. In the current usage pattern, if you break something, it will be immediately visible in the tests. In fact, in such a case, even a single end-to-end SQL test won't be successful.

Reply:
These objects may be used by another developer in a different order than you intend. To make sure that you do not allow them to be performed in the wrong order, it would be good to check all these orders.